### PR TITLE
app: create craft-application based entrypoint.

### PR DESCRIPTION
**Pull request by lengau**

### CODE REVIEW

**Comment by lengau** [comment]
I'll fix the linting and the Windows issue shortly, but I'm focusing on finishing the tests for the pack command

File: requirements-dev.txt [hunk 1]
```diff
@@ -8,6 +8,7 @@ charset-normalizer==3.2.0
 click==8.1.6
 colorama==0.4.6
 coverage==7.3.1
+craft-application@git+https://github.com/canonical/craft-application.git
```
Comment (mr-cal) [inline requirements-dev.txt:11]: Looks like rockcraft is doing this too, so I am fine with this getting merged into a feature branch.

File: charmcraft/application/main.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            "package",
+            work_dir=self._work_dir,
+            prime_dir=self.services.lifecycle.prime_dir,
+        )
+
+    def _get_dispatcher(self) -> craft_cli.Dispatcher:
+        """Configure charmcraft, including a fallback to the classic entrypoint.
+
+        Side-effect: This method may exit the process.
+        Raises: ClassicFallback() to fall back to the classic interface.
```
Comment (mr-cal) [inline charmcraft/application/main.py:57]: Is this style on purpose?  I expected `:raises ClassicFallback: <explanation>`.

**Review by mr-cal — APPROVED** [review]

File: charmcraft/application/main.py (hunk 2 above)
Comment (lengau) [reply charmcraft/application/main.py:57]: Nope! Not intentional. Good catch!

File: requirements-dev.txt (hunk 1 above)
Comment (lengau) [reply requirements-dev.txt:11]: This pulse is probably an appropriate time for a release of craft-application too, which will let us merge this feature branch.

File: charmcraft/application/main.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+        :returns: A ready-to-run Dispatcher object
+        """
+        craft_cli.emit.init(
+            mode=craft_cli.EmitterMode.BRIEF,
+            appname=self.app.name,
+            greeting=f"Starting {self.app.name}",
+            log_filepath=self.log_path,
```
Comment (sergiusens) [inline charmcraft/application/main.py:65]: do you not want to enable streaming mode for brief?

Comment (lengau) [reply charmcraft/application/main.py:65]: Good point! I'll make another PR with that and any other comments :-)

**Pull request merged**